{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "fb7eadbd-7adf-4c53-ac07-3c2dec095ff1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this is a parameter that will get overwritten when run by papermill on a schedules\n",
    "is_local_development = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "40c27b8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: gitlabdata in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (0.3.28)\n",
      "Requirement already satisfied: snowflake-sqlalchemy>=1.1.10 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from gitlabdata) (1.4.4)\n",
      "Requirement already satisfied: pandas>=0.25.3 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from gitlabdata) (1.4.2)\n",
      "Requirement already satisfied: pyyaml==6.0.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from gitlabdata) (6.0)\n",
      "Requirement already satisfied: pygsheets==2.0.5 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from gitlabdata) (2.0.5)\n",
      "Requirement already satisfied: google-api-python-client>=1.5.5 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from pygsheets==2.0.5->gitlabdata) (2.92.0)\n",
      "Requirement already satisfied: google-auth-oauthlib in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from pygsheets==2.0.5->gitlabdata) (1.0.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from pandas>=0.25.3->gitlabdata) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from pandas>=0.25.3->gitlabdata) (2023.3)\n",
      "Requirement already satisfied: numpy>=1.18.5 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from pandas>=0.25.3->gitlabdata) (1.23.5)\n",
      "Requirement already satisfied: sqlalchemy<2.0.0,>=1.4.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from snowflake-sqlalchemy>=1.1.10->gitlabdata) (1.4.48)\n",
      "Requirement already satisfied: snowflake-connector-python<3.0.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from snowflake-sqlalchemy>=1.1.10->gitlabdata) (2.7.8)\n",
      "Requirement already satisfied: httplib2<1.dev0,>=0.15.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from google-api-python-client>=1.5.5->pygsheets==2.0.5->gitlabdata) (0.15.0)\n",
      "Requirement already satisfied: google-auth<3.0.0.dev0,>=1.19.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from google-api-python-client>=1.5.5->pygsheets==2.0.5->gitlabdata) (2.19.1)\n",
      "Requirement already satisfied: google-auth-httplib2>=0.1.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from google-api-python-client>=1.5.5->pygsheets==2.0.5->gitlabdata) (0.1.0)\n",
      "Requirement already satisfied: google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0.dev0,>=1.31.5 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from google-api-python-client>=1.5.5->pygsheets==2.0.5->gitlabdata) (2.11.0)\n",
      "Requirement already satisfied: uritemplate<5,>=3.0.1 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from google-api-python-client>=1.5.5->pygsheets==2.0.5->gitlabdata) (4.1.1)\n",
      "Requirement already satisfied: six>=1.5 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from python-dateutil>=2.8.1->pandas>=0.25.3->gitlabdata) (1.16.0)\n",
      "Requirement already satisfied: asn1crypto<2.0.0,>0.24.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from snowflake-connector-python<3.0.0->snowflake-sqlalchemy>=1.1.10->gitlabdata) (1.5.1)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from snowflake-connector-python<3.0.0->snowflake-sqlalchemy>=1.1.10->gitlabdata) (2023.5.7)\n",
      "Requirement already satisfied: cffi<2.0.0,>=1.9 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from snowflake-connector-python<3.0.0->snowflake-sqlalchemy>=1.1.10->gitlabdata) (1.15.1)\n",
      "Requirement already satisfied: charset-normalizer~=2.0.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from snowflake-connector-python<3.0.0->snowflake-sqlalchemy>=1.1.10->gitlabdata) (2.0.12)\n",
      "Requirement already satisfied: cryptography<37.0.0,>=3.1.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from snowflake-connector-python<3.0.0->snowflake-sqlalchemy>=1.1.10->gitlabdata) (36.0.2)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from snowflake-connector-python<3.0.0->snowflake-sqlalchemy>=1.1.10->gitlabdata) (3.4)\n",
      "Requirement already satisfied: oscrypto<2.0.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from snowflake-connector-python<3.0.0->snowflake-sqlalchemy>=1.1.10->gitlabdata) (1.3.0)\n",
      "Requirement already satisfied: pyOpenSSL<23.0.0,>=16.2.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from snowflake-connector-python<3.0.0->snowflake-sqlalchemy>=1.1.10->gitlabdata) (22.0.0)\n",
      "Requirement already satisfied: pycryptodomex!=3.5.0,<4.0.0,>=3.2 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from snowflake-connector-python<3.0.0->snowflake-sqlalchemy>=1.1.10->gitlabdata) (3.18.0)\n",
      "Requirement already satisfied: pyjwt<3.0.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from snowflake-connector-python<3.0.0->snowflake-sqlalchemy>=1.1.10->gitlabdata) (2.7.0)\n",
      "Requirement already satisfied: requests<3.0.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from snowflake-connector-python<3.0.0->snowflake-sqlalchemy>=1.1.10->gitlabdata) (2.31.0)\n",
      "Requirement already satisfied: setuptools>34.0.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from snowflake-connector-python<3.0.0->snowflake-sqlalchemy>=1.1.10->gitlabdata) (67.8.0)\n",
      "Requirement already satisfied: greenlet!=0.4.17 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from sqlalchemy<2.0.0,>=1.4.0->snowflake-sqlalchemy>=1.1.10->gitlabdata) (2.0.2)\n",
      "Requirement already satisfied: requests-oauthlib>=0.7.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from google-auth-oauthlib->pygsheets==2.0.5->gitlabdata) (1.3.1)\n",
      "Requirement already satisfied: pycparser in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from cffi<2.0.0,>=1.9->snowflake-connector-python<3.0.0->snowflake-sqlalchemy>=1.1.10->gitlabdata) (2.21)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0dev,>=1.56.2 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0.dev0,>=1.31.5->google-api-python-client>=1.5.5->pygsheets==2.0.5->gitlabdata) (1.59.0)\n",
      "Requirement already satisfied: protobuf!=3.20.0,!=3.20.1,!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.19.5 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0.dev0,>=1.31.5->google-api-python-client>=1.5.5->pygsheets==2.0.5->gitlabdata) (3.19.6)\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from google-auth<3.0.0.dev0,>=1.19.0->google-api-python-client>=1.5.5->pygsheets==2.0.5->gitlabdata) (5.3.1)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from google-auth<3.0.0.dev0,>=1.19.0->google-api-python-client>=1.5.5->pygsheets==2.0.5->gitlabdata) (0.3.0)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from google-auth<3.0.0.dev0,>=1.19.0->google-api-python-client>=1.5.5->pygsheets==2.0.5->gitlabdata) (4.9)\n",
      "Requirement already satisfied: urllib3<2.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from google-auth<3.0.0.dev0,>=1.19.0->google-api-python-client>=1.5.5->pygsheets==2.0.5->gitlabdata) (1.26.16)\n",
      "Requirement already satisfied: oauthlib>=3.0.0 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib->pygsheets==2.0.5->gitlabdata) (3.2.2)\n",
      "Requirement already satisfied: pyasn1<0.6.0,>=0.4.6 in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from pyasn1-modules>=0.2.1->google-auth<3.0.0.dev0,>=1.19.0->google-api-python-client>=1.5.5->pygsheets==2.0.5->gitlabdata) (0.5.0)\n"
     ]
    }
   ],
   "source": [
    "!python -m pip install gitlabdata --upgrade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "f1aa4a4d-fe9d-447f-84c7-d0f88c340bb6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pyprojroot in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (0.3.0)\n",
      "Requirement already satisfied: typing-extensions in /Users/nfiguera/.local/share/virtualenvs/data-science-mn1T8veA/lib/python3.9/site-packages (from pyprojroot) (4.2.0)\n"
     ]
    }
   ],
   "source": [
    "!python -m pip install pyprojroot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "edea41d9-b0a0-415d-998a-9e7c0f3b8312",
   "metadata": {},
   "outputs": [],
   "source": [
    "import configparser\n",
    "\n",
    "# import sys\n",
    "import pandas as pd\n",
    "from datetime import datetime\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from datetime import date\n",
    "import json, os\n",
    "from pyprojroot import here\n",
    "from os import environ as env"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "4ca065e8-5fc5-409a-a428-5eb2d9e3ad72",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gitlabdata.orchestration_utils import (\n",
    "    data_science_engine_factory,\n",
    "    query_dataframe,\n",
    "    snowflake_engine_factory,\n",
    "    snowflake_stage_load_copy_remove,\n",
    "    get_env_from_profile,\n",
    "    dataframe_uploader,\n",
    "    write_to_gsheets,\n",
    "    read_from_gsheets,\n",
    "    query_executor,\n",
    "    query_from_file,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a898e93-3838-4401-842e-320b4336557a",
   "metadata": {},
   "source": [
    "## Create Snowflake engine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "859777a1-99e7-42a1-a2eb-a895ba919c32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Engine(snowflake://nfiguera%40gitlab.com:***@gitlab/RAW/?authenticator=externalbrowser&role=NFIGUERA&warehouse=DEV_XS)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# engine factory can be created using a local role from output\n",
    "# depending on this notebook being run locally or remotely, the\n",
    "# engine is creation process is different\n",
    "\n",
    "if is_local_development:\n",
    "    snowflake_engine = data_science_engine_factory(\n",
    "        profile_target=\"sales_analytics_local\"\n",
    "    )\n",
    "else:\n",
    "    snowflake_engine = snowflake_engine_factory(env, \"SALES_ANALYTICS\")\n",
    "\n",
    "    raw_db_name = env[\"SNOWFLAKE_LOAD_DATABASE\"]\n",
    "    prod_db_name = env[\"SNOWFLAKE_PROD_DATABASE\"]\n",
    "\n",
    "snowflake_engine"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4516078d-439b-4276-bef1-53ace391225a",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Credentials for Gsheet manipulation\n",
    "\n",
    "Remember to give access to the following two users:\n",
    "\n",
    "- Data Team runner: data-team-sheets-sa@gitlab-analysis.iam.gserviceaccount.com\n",
    "- Sales Strategy service account: service-revenue-strat-analytic@revenue-strategy-anal-411d5a72.iam.gserviceaccount.com"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "54ec218f-85cf-4ec3-92e2-012799c81475",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the credentials of the google service account\n",
    "if is_local_development:\n",
    "    credentials_path = here(\"credentials/rsa_gcloud_service_account.json\")\n",
    "    # credentials_path = here(\"credentials/gsheet_service_file.json\")\n",
    "\n",
    "    with open(credentials_path) as f:\n",
    "        service_account_credentials = f.read().replace(\"\\n\", \"\")\n",
    "\n",
    "    # set the credential as a enviroment variable\n",
    "    os.environ[\"GSHEETS_SERVICE_ACCOUNT_CREDENTIALS\"] = service_account_credentials"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eccb2b25-72c0-4b5d-812e-59cbcf16bc40",
   "metadata": {},
   "source": [
    "# SAL Heatmap daily ETL update process\n",
    "\n",
    "This notebook is run daily to update the SAL Heatmap source file. \n",
    "\n",
    "The process is responsible of updating the following datasets:\n",
    "\n",
    "1. QTD SAL Extracts\n",
    "2. Quotas\n",
    "3. Regional Hist Coverages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "5f865d91-3b91-45d1-9379-85497607e07f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SOURCE SAL Heatmap Sheet ID\n",
    "SOURCE_GSHEET_ID = \"1Ufh2DKwTq25IHzZeaCo0bFyQcFNvCtC3bVnR3oOZz-g\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81b356fa-3547-4608-a852-52ff741c6e91",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Excute Snowflake query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "5d611210",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initiating login request with your identity provider. A browser window should have opened for you to complete the login. If you can't see it, check existing browser windows, or your OS settings. Press CTRL+C to abort and try again...\n"
     ]
    }
   ],
   "source": [
    "# QTD metrics\n",
    "qtd_metrics = query_from_file(snowflake_engine, \"sh_qtd_metrics_for_sal_heatmap.sql\")\n",
    "qtd_metrics[\"last_extracted_etl\"] = date.today()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "bc9bfa9f-0120-4c34-af91-0d03861ac582",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20591713.339999996"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Test\n",
    "qtd_metrics.qtd_booked_net_arr.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "eaa7a645-677e-4841-a79b-fe5c9366f640",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initiating login request with your identity provider. A browser window should have opened for you to complete the login. If you can't see it, check existing browser windows, or your OS settings. Press CTRL+C to abort and try again...\n"
     ]
    }
   ],
   "source": [
    "# Historical coverages\n",
    "hist_coverage = query_from_file(\n",
    "    snowflake_engine, \"sh_hist_qtr_coverages_for_sal_heatmap.sql\"\n",
    ")\n",
    "hist_coverage[\"last_extracted_etl\"] = date.today()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "040b9299-d65a-4bad-81d9-78ba369f17ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initiating login request with your identity provider. A browser window should have opened for you to complete the login. If you can't see it, check existing browser windows, or your OS settings. Press CTRL+C to abort and try again...\n"
     ]
    }
   ],
   "source": [
    "# Quotas\n",
    "# NF 20230531 The Xactly version of the quota capture doesn't work\n",
    "# potentially data is not correct in Xactly\n",
    "# fy24_quota = query_from_file(snowflake_engine, \"sh_xactly_fy24_quotas.sql\")\n",
    "\n",
    "# NF 20230531 The table in this query is uploaded from a raw gsheet maintained by Haseem and\n",
    "# uploaded into the raw sales analytics table\n",
    "fy24_quota = query_from_file(snowflake_engine, \"sh_raw_fy24_quotas.sql\")\n",
    "fy24_quota[\"last_extracted_etl\"] = date.today()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "59c642f7-3d92-474e-b534-024b3302167d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "249978101"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Test\n",
    "fy24_quota.cfy_total.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "b4cc29b6-abf0-4f9d-9fb7-4f9e4bc16439",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initiating login request with your identity provider. A browser window should have opened for you to complete the login. If you can't see it, check existing browser windows, or your OS settings. Press CTRL+C to abort and try again...\n"
     ]
    }
   ],
   "source": [
    "# Employee ID\n",
    "empl_list = query_from_file(snowflake_engine, \"sh_empid_list.sql\")\n",
    "empl_list[\"last_extracted_etl\"] = date.today()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eed27c40-f8c9-47ab-990d-ef7f47d2b0a6",
   "metadata": {},
   "source": [
    "## QTD Metrics Extracts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52025c13-f77f-49b4-81e8-bfbd26eb48f9",
   "metadata": {},
   "source": [
    "### Write to SAL HEatmap source gSheet file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "f723d4fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:<HttpError 400 when requesting https://sheets.googleapis.com/v4/spreadsheets/1Ufh2DKwTq25IHzZeaCo0bFyQcFNvCtC3bVnR3oOZz-g:batchUpdate?fields=replies%2FaddSheet&alt=json returned \"Invalid requests[0].addSheet: A sheet with the name \"jupyter_qtd_sal\" already exists. Please enter another name.\". Details: \"Invalid requests[0].addSheet: A sheet with the name \"jupyter_qtd_sal\" already exists. Please enter another name.\">\n"
     ]
    }
   ],
   "source": [
    "# Write to GSheets\n",
    "sheet_id = SOURCE_GSHEET_ID\n",
    "sheet_name = \"jupyter_qtd_sal\"\n",
    "qtd_metrics.columns = qtd_metrics.columns.str.upper()\n",
    "write_to_gsheets(sheet_id, sheet_name, qtd_metrics)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cab7f41f-67d8-4c0b-ba15-47443beae471",
   "metadata": {},
   "source": [
    "## Historical QTR Coverages Data Extract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "81e188a3-851f-4837-967d-3fb30ec8c754",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:<HttpError 400 when requesting https://sheets.googleapis.com/v4/spreadsheets/1Ufh2DKwTq25IHzZeaCo0bFyQcFNvCtC3bVnR3oOZz-g:batchUpdate?fields=replies%2FaddSheet&alt=json returned \"Invalid requests[0].addSheet: A sheet with the name \"jupyter_hist_qtr_aggregated\" already exists. Please enter another name.\". Details: \"Invalid requests[0].addSheet: A sheet with the name \"jupyter_hist_qtr_aggregated\" already exists. Please enter another name.\">\n"
     ]
    }
   ],
   "source": [
    "# Write to GSheets\n",
    "sheet_id = SOURCE_GSHEET_ID\n",
    "sheet_name = \"jupyter_hist_qtr_aggregated\"\n",
    "hist_coverage.columns = hist_coverage.columns.str.upper()\n",
    "write_to_gsheets(sheet_id, sheet_name, hist_coverage)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e27cb822-bb71-40c4-85ea-fc385f59917d",
   "metadata": {},
   "source": [
    "## FY24 Quota"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "5820cf49-bfa5-43ab-ae0c-b85b9d13f349",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:<HttpError 400 when requesting https://sheets.googleapis.com/v4/spreadsheets/1Pa3bh2wp1mPCtln0_3mH-r627IEJA5yofAFAir_kvBo:batchUpdate?fields=replies%2FaddSheet&alt=json returned \"Invalid requests[0].addSheet: A sheet with the name \"jupyter_fy_quota\" already exists. Please enter another name.\". Details: \"Invalid requests[0].addSheet: A sheet with the name \"jupyter_fy_quota\" already exists. Please enter another name.\">\n"
     ]
    }
   ],
   "source": [
    "# Write to GSheets\n",
    "sheet_id = \"1Pa3bh2wp1mPCtln0_3mH-r627IEJA5yofAFAir_kvBo\"\n",
    "sheet_name = \"jupyter_fy_quota\"\n",
    "write_to_gsheets(sheet_id, sheet_name, fy24_quota)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e0d810a-16f7-446b-88a5-c5a98027dd22",
   "metadata": {},
   "source": [
    "## Employee List "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "d951d06f-c9ce-4ce4-a02a-a6b5b37464d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:<HttpError 400 when requesting https://sheets.googleapis.com/v4/spreadsheets/1Ufh2DKwTq25IHzZeaCo0bFyQcFNvCtC3bVnR3oOZz-g:batchUpdate?fields=replies%2FaddSheet&alt=json returned \"Invalid requests[0].addSheet: A sheet with the name \"jupyter_empl_id\" already exists. Please enter another name.\". Details: \"Invalid requests[0].addSheet: A sheet with the name \"jupyter_empl_id\" already exists. Please enter another name.\">\n"
     ]
    }
   ],
   "source": [
    "# Write to GSheets\n",
    "sheet_id = SOURCE_GSHEET_ID\n",
    "sheet_name = \"jupyter_empl_id\"\n",
    "write_to_gsheets(sheet_id, sheet_name, empl_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9d61acd-b099-4ffc-853d-624524ddee01",
   "metadata": {},
   "source": [
    "## Clari Forecast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "3714354c-9fca-425d-9a3c-ef870bda31da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initiating login request with your identity provider. A browser window should have opened for you to complete the login. If you can't see it, check existing browser windows, or your OS settings. Press CTRL+C to abort and try again...\n"
     ]
    }
   ],
   "source": [
    "# Execute clari forecast\n",
    "clari_forecast = query_from_file(snowflake_engine, \"sh_forecast_update.sql\")\n",
    "clari_forecast[\"last_extracted_etl\"] = date.today()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "b9f799e8-0e97-40a9-b6e5-bd876339b79b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:<HttpError 400 when requesting https://sheets.googleapis.com/v4/spreadsheets/1Ufh2DKwTq25IHzZeaCo0bFyQcFNvCtC3bVnR3oOZz-g:batchUpdate?fields=replies%2FaddSheet&alt=json returned \"Invalid requests[0].addSheet: A sheet with the name \"jupyter_clari_forecast\" already exists. Please enter another name.\". Details: \"Invalid requests[0].addSheet: A sheet with the name \"jupyter_clari_forecast\" already exists. Please enter another name.\">\n"
     ]
    }
   ],
   "source": [
    "# Write to GSheets\n",
    "sheet_id = SOURCE_GSHEET_ID\n",
    "sheet_name = \"jupyter_clari_forecast\"\n",
    "write_to_gsheets(sheet_id, sheet_name, clari_forecast)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4d58cb9-e769-4515-9e26-d5a9daccbbc6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
